brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                1.to(repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
-    }
-    copy(trees = newTrees.toList)
+trait TrainingStep[K,V,T] {
+
+  type K1
+  type V1
+  type V2
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]): Seq[((Int,Int,K1), V1)]
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: V1): Traversable[V2]
+  def update(tree: Tree[K,V,T], map: Map[Int,V2]): Tree[K,V,T]
+
+  def semigroup1: Semigroup[V1]
+  def semigroup2: Semigroup[V2]
+  def ordering: Ordering[K1]
+}
+
+case class UpdateTargets[K,V,T](sampler: Sampler[K])(implicit val semigroup1: Semigroup[T])
+  extends TrainingStep[K,V,T] {
+
+  type K1 = Unit
+  type V1 = T
+  type V2 = T
+
+  val semigroup2 = semigroup1
+  val ordering = implicitly[Ordering[Unit]]
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+     for (
+        (treeIndex, tree) <- trees.toList;
+        i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+        leafIndex <- tree.leafIndexFor(instance.features).toList
+      ) yield (treeIndex, leafIndex, ()) -> instance.target
   }
 
-  private def updateLeaves(fn: (Int, LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]) => Node[K, V, T, Unit]): Trainer[K, V, T] = {
-    updateTrees {
-      case (tree, treeIndex, byLeaf) =>
-        val newNodes = byLeaf.map {
-          case (leaf, instances) =>
-            leaf.index -> fn(treeIndex, leaf, instances)
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: T) = List(v1)
+
+  def update(tree: Tree[K,V,T], map: Map[Int,T]) = {
+    tree.updateByLeafIndex { index => map.get(index).map { t => LeafNode(index, t) } }
+  }
+}
+
+case class Expand[K,V,T](sampler: Sampler[K], stopper: Stopper[T], splitter: Splitter[V, T], evaluator: Evaluator[V, T])(implicit val ordering: Ordering[K])
+  extends TrainingStep[K,V,T] {
+
+  type K1 = K
+  type V1 = splitter.S
+  type V2 = (K, Split[V, T], Double)
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+    lazy val features = instance.features.mapValues { value => splitter.create(value, instance.target) }
+
+    for
+      ((treeIndex, tree) <- trees.toList;
+      i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+      leaf <- tree.leafFor(instance.features).toList if stopper.shouldSplit(leaf.target);
+      (feature, stats) <- features if (sampler.includeFeature(feature, treeIndex, leaf.index)))
+        yield (treeIndex, leaf.index, feature) -> stats
+  }
+
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K, v1: V1) = {
+    tree.leafAt(leafIndex).toList.flatMap { leaf =>
```

Comment:
Your inverted tree idea would let us get rid of this full-tree traversal, right?